classes/payment_manager.php

```diff
@@ -0,0 +1,126 @@
+    const PAYMENT_COMPLETE = 1;
+    const PAYMENT_INCOMPLETE = 2;
+
+    /**
+     * Enrols a user in a course.
+     *
+     * @param int $courseid
+     * @param int $userid
+     * @throws \moodle_exception
+     */
+    public static function paymentplugin_enrol(int $courseid, int $userid) {
+        global $DB;
+        if (!$DB->record_exists('course', array('id' => $courseid))) {
+            throw new \moodle_exception('errorinvalidcourse', 'tool_paymentplugin', '', $courseid);
+        }
+
+        if (!$DB->record_exists('user', array('id' => $userid))) {
+            throw new \moodle_exception('errorinvaliduser', 'tool_paymentplugin', '', $userid);
+        }
+
+        $enrolinstance = $DB->get_record('enrol', array('enrol' => 'payment', 'courseid' => $courseid));
+        if (!$enrolinstance) {
+            throw new \moodle_exception('errorinvalidcourseenrol', 'tool_paymentplugin', '', $courseid);
+        }
+
+        $enrol = enrol_get_plugin('payment');
+        $enrol->enrol_user($enrolinstance, $userid);
+    }
+
+    public static function filter_underscores($data) : \stdClass {
+        $newdata = new \stdclass();
+        foreach ($data as $var => $value) {
+            $var = str_replace('_', '', $var);
+            $newdata->$var = $value;
+        }
+        return $newdata;
+    }
+
+    /**
+     * Actions a transaction given the correct data.
+     *
+     * @param stdClass $gateway A payment gateway instance.
+     * @param int $paymentstatus Either PAYMENT_FAILED, PAYMENT_COMPLETE or PAYMENT_INCOMPLETE.
+     * @param int $userid The moodle id of the user making the purchase.
+     * @param string $currency The currency the transaction was made in.
+     * @param float $amount the value of the amount paid.
+     * @param int $date The date time of the purchase.
+     * @param int $courseid The moodle course id that the transaction was used to purchase.
+     * @param \stdclass $additionaldata Any valid additional data in this object will be inserted into the
+     * specified table $gatewaytablename.
+     */
+    public static function submit_transaction(\tool_paymentplugin\paymentgateway\object_paymentgateway $gateway,
+        int $paymentstatus, int $userid, string $currency, float $amount, int $date, int $courseid,
+        array $additionaldata = null) : int {
+        global $DB;
+
+        // Ensure currency is a valid currency.
+        if (!in_array(strtolower($currency), self::CURRENCY)) {
+            throw new \moodle_exception('Invalid currency passed.');
+        }
+
+        $gatewayname = $gateway->get_name();
+        $gatewaytablename = $gateway->get_tablename();
+
+        $id = $DB->insert_record('tool_paymentplugin_purchases', ['gateway' => $gatewayname, 'currency' => $currency,
+            'userid' => $userid, 'amount' => $amount, 'paymentdate' => $date, 'courseid' => $courseid, 'success' => $paymentstatus]
+            );
+
+        if (!is_null($additionaldata)) {
```

Comment:
This should perhaps be an empty() check instead of is_null, incase someone passes an empty array, which technically isnt null, but shouldnt be operated on.